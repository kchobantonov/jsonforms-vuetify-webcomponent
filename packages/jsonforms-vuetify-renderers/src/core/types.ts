import type {
  JsonFormsCellRendererRegistryEntry,
  JsonFormsI18nState,
  JsonFormsRendererRegistryEntry,
  JsonFormsSubStates,
  JsonFormsUISchemaRegistryEntry,
  JsonSchema,
  Middleware,
  Translator,
  UISchemaElement,
  ValidationMode,
} from '@jsonforms/core';
import type { MaybeReadonly } from '@jsonforms/vue';
import type { Ajv, ErrorObject } from 'ajv';
import type {
  Component,
  ComputedOptions,
  ComputedRef,
  Directive,
  InjectionKey,
  MethodOptions,
  Ref,
  SetupContext,
} from 'vue';

export const AsyncFunction = Object.getPrototypeOf(async function (
  _event: ActionEvent,
) {}).constructor;

export interface JsonFormsProps {
  data: any;
  schema?: JsonSchema;
  schemaUrl?: string;
  uischema?: UISchemaElement;
  renderers: MaybeReadonly<JsonFormsRendererRegistryEntry[]>;
  cells?: MaybeReadonly<JsonFormsCellRendererRegistryEntry[]>;
  config?: any;
  readonly?: boolean;
  uischemas?: MaybeReadonly<JsonFormsUISchemaRegistryEntry[]>;
  validationMode?: ValidationMode;
  ajv?: Ajv;
  i18n?: JsonFormsI18nState & { translations?: Record<string, any> };
  additionalErrors?: ErrorObject<string, Record<string, any>, unknown>[];
  middleware?: Middleware;
  onHandleAction?: (event: ActionEvent) => void;
}

export type ResolvedSchema = {
  schema?: JsonSchema;
  resolved: boolean;
  error?: string;
};

export interface FormContext {
  [key: string]: any;

  schemaUrl?: string;

  jsonforms?: JsonFormsSubStates;

  // below are just the shortcuts for acessing the jsonforms.core
  locale?: string;
  translate?: Translator;
  data?: any;
  schema?: JsonSchema;
  uischema?: UISchemaElement;
  errors?: ErrorObject[];
  additionalErrors?: ErrorObject[];
}

export type ActionEvent = {
  action: string;
  callback?: (event: ActionEvent) => void;
  jsonforms: JsonFormsSubStates;
  context: FormContext;
  // the action parameters passes from the UI schema
  params: Record<string, any>;
  $el: Element;
};

export interface NamedUISchemaElement extends UISchemaElement {
  name: string;
}

export interface DataProvider {
  loading: boolean;
  data: any;
  error: any;
}

export const DataProviderKey: InjectionKey<DataProvider> = Symbol.for(
  'jsonforms-vuetify-renderers:dataProvider',
);

export const TemplateDirectivesKey: InjectionKey<Record<string, Directive>> =
  Symbol.for(
    'jsonforms-vuetify-renderers:templateLayoutRendererComponentDirectives',
  );

export const TemplateComputedKey: InjectionKey<ComputedOptions> = Symbol.for(
  'jsonforms-vuetify-renderers:templateLayoutRendererComponentComputed',
);

export const TemplateMethodsKey: InjectionKey<MethodOptions> = Symbol.for(
  'jsonforms-vuetify-renderers:templateLayoutRendererComponentMethods',
);

export const TemplateFiltersKey: InjectionKey<MethodOptions> = Symbol.for(
  'jsonforms-vuetify-renderers:templateLayoutRendererComponentFilters',
);

export const TemplateComponentsKey: InjectionKey<Record<string, Component>> =
  Symbol.for(
    'jsonforms-vuetify-renderers:templateLayoutRendererComponentComponents',
  );

export const FormContextKey: InjectionKey<Ref<FormContext>> = Symbol.for(
  'jsonforms-vuetify-renderers:formContext',
);

export const TemplateRenderSlotContentsKey: InjectionKey<
  ComputedRef<Record<string, UISchemaElement>>
> = Symbol.for('jsonforms-vuetify-renderers:templateRendererSlotContents');

export const HandleActionEmitterKey: InjectionKey<SetupContext['emit']> =
  Symbol.for('jsonforms-vuetify-renderers:handleActionEmitter');
