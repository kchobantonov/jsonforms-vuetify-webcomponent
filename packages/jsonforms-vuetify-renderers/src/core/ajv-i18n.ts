import type { Ajv, Schema, ValidateFunction } from 'ajv';
import localize from 'ajv-i18n/localize';
import type {
  AnySchema,
  AnyValidateFunction,
  AsyncSchema,
  AsyncValidateFunction,
  JSONSchemaType,
  JTDDataType,
  JTDSchemaType,
  SomeJTDSchemaType,
} from 'ajv/dist/core';
import { unref, type ComputedRef, type MaybeRef } from 'vue';

export const enableErrorTranslations = (
  ajv: Ajv,
  locale: MaybeRef<string | undefined> | ComputedRef<string | undefined>,
) => {
  const originalCompile = ajv.compile.bind(ajv);

  function compile<T = unknown>(
    schema: Schema | JSONSchemaType<T>,
    _meta?: boolean,
  ): ValidateFunction<T>;
  // Separated for type inference to work
  // eslint-disable-next-line @typescript-eslint/unified-signatures
  function compile<T = unknown>(
    schema: JTDSchemaType<T>,
    _meta?: boolean,
  ): ValidateFunction<T>;
  // This overload is only intended for typescript inference, the first
  // argument prevents manual type annotation from matching this overload
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  function compile<N extends never, T extends SomeJTDSchemaType>(
    schema: T,
    _meta?: boolean,
  ): ValidateFunction<JTDDataType<T>>;
  function compile<T = unknown>(
    schema: AsyncSchema,
    _meta?: boolean,
  ): AsyncValidateFunction<T>;
  function compile<T = unknown>(
    schema: AnySchema,
    _meta?: boolean,
  ): AnyValidateFunction<T>;
  function compile<T = unknown>(
    schema: AnySchema,
    _meta?: boolean,
  ): AnyValidateFunction<T> {
    const originalValidate = originalCompile(schema, _meta);

    const makeValidate = () => {
      // Create a new function using a Proxy to intercept property access and modifications
      const validate = ((...args: Parameters<ValidateFunction<T>>) => {
        const isValid = originalValidate(...args); // Call the original function

        if (!isValid && originalValidate.errors) {
          const language = unref(locale) as keyof typeof localize | undefined;
          const localized =
            language &&
            (localize[language] ||
              localize[language.split('-')[0] as keyof typeof localize]);
          if (localized) localized(originalValidate.errors);
        }
        return isValid;
      }) as ValidateFunction<T>;

      // Create a Proxy to synchronize properties between validate and originalValidate
      const handler = {
        get(target: any, prop: string, receiver: any) {
          if (prop in target) {
            return target[prop];
          }
          return (originalValidate as any)[prop];
        },
        set(target: any, prop: string, value: any) {
          // Sync properties between validate and originalValidate
          if (prop in target) {
            target[prop] = value;
          }
          (originalValidate as any)[prop] = value;
          return true;
        },
      };

      // Wrap the validate function in a Proxy to make sure properties are synchronized
      const proxyValidate = new Proxy(validate, handler);

      return proxyValidate;
    };

    return makeValidate();
  }

  ajv.compile = compile;
  return ajv;
};
