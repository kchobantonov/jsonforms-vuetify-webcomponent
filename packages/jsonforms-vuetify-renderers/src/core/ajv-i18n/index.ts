import type { Ajv, ErrorObject, Plugin } from 'ajv';
import localize from 'ajv-i18n/localize';
import { unref, type ComputedRef, type MaybeRef } from 'vue';
import { localize_bg } from './bg';
import type { JsonFormsI18nState, Translator } from '@jsonforms/core';

function injectAdditionalLocales() {
  // add missing BG locale - https://github.com/ajv-validator/ajv-i18n/pull/312 until this is merged and released
  const bg = 'bg' as keyof typeof localize;
  if (!localize[bg]) {
    localize[bg] = localize_bg;
  }

  // override the required message so that the missingProperty is not displayed as not translated text
  localize[bg] = (errors) => {
    if (!(errors && errors.length)) return;

    localize_bg(errors);
    for (const e of errors) {
      if (e.keyword === 'required') {
        e.message = 'полето е задължително';
      }
    }
  };
}

export interface I18NOptions {
  i18n:
    | MaybeRef<JsonFormsI18nState | undefined>
    | ComputedRef<JsonFormsI18nState | undefined>;
}

export const ajvTranslations: Plugin<I18NOptions> = (
  ajv: Ajv,
  options?: I18NOptions,
): Ajv => {
  injectAdditionalLocales();

  const compileFunction = ajv.compile;
  ajv.compile = Object.assign(
    (...compileProps: Parameters<typeof compileFunction>) => {
      const validateFunction = compileFunction.apply(ajv, compileProps);

      const newValidateFunction = Object.assign(
        (...innerProps: Parameters<typeof validateFunction>) => {
          const validationResult = validateFunction.apply(
            compileFunction,
            innerProps,
          );

          if (!validationResult && validateFunction.errors) {
            const i18n = options?.i18n ? unref(options.i18n) : undefined;
            const language: keyof typeof localize | undefined = i18n
              ? (i18n.locale as keyof typeof localize)
              : undefined;

            const localized =
              language &&
              (localize[language] ||
                localize[language.split('-')[0] as keyof typeof localize]);
            if (localized) {
              localized(
                // exclude errorMessage keyword since the message itself will be comming from the schema itself and possible then translated
                validateFunction.errors.filter(
                  (error) => error.keyword !== 'errorMessage',
                ),
              );
            }

            validateFunction.errors = unwrapErrorMessageErrors(
              validateFunction.errors,
              i18n?.translate,
            );
          }
          Object.assign(newValidateFunction, validateFunction);
          return validationResult;
        },
        validateFunction,
      );

      return newValidateFunction;
    },
    compileFunction,
  );

  return ajv;
};

export const unwrapErrorMessageErrors = (
  errors: ErrorObject[],
  translate: Translator | undefined,
): ErrorObject[] => {
  return (
    errors
      // filter out any errors that are marked as processed by ajv-errors
      .filter((error) => (error as any)['emUsed'] !== true)
      .map((error) => {
        if (
          error.keyword === 'errorMessage' &&
          Array.isArray(error.params?.errors)
        ) {
          // error created by the ajv-errors, replace the error with the actual errors and change the error message
          return error.params.errors.map((paramError) => ({
            ...paramError,
            message:
              translate && error.message
                ? translate(error.message, error.message)
                : error.message,
          }));
        }
        // otherwise just return the current error that is not related to ajv-errors
        return error;
      })
      .flat()
  );
};
