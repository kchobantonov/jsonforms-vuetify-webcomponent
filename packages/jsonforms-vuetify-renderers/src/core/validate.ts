import type { Options } from 'ajv';
import { createAjv as createDefaultAjv } from '@jsonforms/vue-vuetify';
import { ajvKeywords } from './keywords';

export const createAjv = () => {
  const options: Options = {
    useDefaults: true,
    $data: true,
    discriminator: true,
  };

  const ajv = createDefaultAjv(options);
  ajvKeywords(ajv);

  return ajv;
};
