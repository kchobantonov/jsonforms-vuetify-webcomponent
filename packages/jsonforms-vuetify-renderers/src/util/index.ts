export * from './composition';
export * from './util';
