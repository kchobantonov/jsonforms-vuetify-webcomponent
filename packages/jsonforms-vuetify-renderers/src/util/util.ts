import {
  type JsonFormsUISchemaRegistryEntry,
  type UISchemaElement,
  type UISchemaTester,
  NOT_APPLICABLE,
} from '@jsonforms/core';

export const parseAndTransformUISchemaRegistryEntries = (
  uischemaRegistryEntries?: string,
): JsonFormsUISchemaRegistryEntry[] => {
  const uischemasMap: {
    tester: string;
    uischema: UISchemaElement;
  }[] =
    typeof uischemaRegistryEntries === 'string'
      ? JSON.parse(uischemaRegistryEntries)
      : [];

  return uischemasMap
    .map((elem, index) => {
      if (elem.tester) {
        const action: UISchemaTester = (jsonSchema, schemaPath, path) => {
          try {
            const tester = new Function(
              'jsonSchema, schemaPath, path',
              `const NOT_APPLICABLE = -1; const tester = ${elem.tester}; return tester(jsonSchema, schemaPath, path);`,
            );
            const result = tester(jsonSchema, schemaPath, path);
            if (typeof result !== 'number') {
              console.error(
                `Error at uischema tester[${index}]: invalid result type, expected number but got ${typeof result}`,
              );
            }
            return typeof result === 'number' ? result : NOT_APPLICABLE;
          } catch (e) {
            console.error(`Error at uischema tester[${index}]: ${e}`);
            return NOT_APPLICABLE;
          }
        };
        return {
          tester: action,
          uischema: elem.uischema,
        };
      }
      return null;
    })
    .filter((x) => !!x) as JsonFormsUISchemaRegistryEntry[];
};
