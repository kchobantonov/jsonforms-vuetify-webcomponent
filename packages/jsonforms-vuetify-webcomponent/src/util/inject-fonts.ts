import * as shadyCss from 'shady-css-parser';

class FontFaceExtractor extends shadyCss.Stringifier {
  insideFontFace = false;
  visit(node: shadyCss.Node): string | undefined {
    if (node.type === shadyCss.nodeType.stylesheet) {
      return super.visit(node);
    }
    if (node.type === shadyCss.nodeType.atRule && node.name === 'font-face') {
      try {
        this.insideFontFace = true;
        return super.visit(node);
      } finally {
        this.insideFontFace = false;
      }
    }
    return this.insideFontFace ? super.visit(node) : '';
  }
}

class ExtractImportUrls extends shadyCss.Stringifier {
  importUrls: string[] = [];

  visit(node: shadyCss.Node): string | undefined {
    if (node.type === shadyCss.nodeType.atRule && node.name === 'import') {
      // Check if the import has a valid URL
      const urlMatch = node.parameters.match(/url\(['"]?([^'"]+)['"]?\)/);
      if (urlMatch && urlMatch[1]) {
        this.importUrls.push(urlMatch[1]);
      }
    }
    return super.visit(node);
  }

  // Function to extract import URLs from the given CSS text
  static extract(css: string): string[] {
    const extractor = new ExtractImportUrls();
    const parser = new shadyCss.Parser();
    const ast = parser.parse(css);

    extractor.visit(ast);
    return extractor.importUrls;
  }
}

function adjustFontFaceUrls(cssContent: string, baseUrl: string): string {
  const fontFaceRegex = /@font-face\s*{[^}]*}/g;
  return cssContent.replace(fontFaceRegex, (fontFaceBlock) => {
    return fontFaceBlock.replace(/url\(['"]?([^'"]+)['"]?\)/g, (match, url) => {
      // Check if the URL is relative
      if (!url.startsWith('http://') && !url.startsWith('https://')) {
        // Construct the new URL relative to the base URL
        const adjustedUrl = new URL(url, baseUrl).href; // Create absolute URL
        return `url('${adjustedUrl}')`; // Return adjusted URL
      }
      return match;
    });
  });
}

const processStyleNode = async (
  styleNode: HTMLStyleElement,
  fontFaceRules: string[],
) => {
  if (!styleNode.textContent) return;
  return processCssText(styleNode.textContent, fontFaceRules);
};

const processCssText = async (cssText: string, fontFaceRules: string[]) => {
  const parser = new shadyCss.Parser();
  const cssAst = parser.parse(cssText);

  const fontFaceExtractor = new FontFaceExtractor();
  const fontFaceCss = fontFaceExtractor.stringify(cssAst);
  if (fontFaceCss) {
    fontFaceRules.push(fontFaceCss);
  }

  const cssUrls = ExtractImportUrls.extract(cssText);
  // check for @import rules and analyzed if they refer to font-faces
  for (const cssUrl of cssUrls) {
    const url = cssUrl.startsWith('/')
      ? window.location.origin + cssUrl
      : cssUrl;

    const response = import.meta.env.DEV
      ? await fetch(url + '?raw')
      : await fetch(url);

    if (response.ok) {
      let importedCssText = await response.text();
      if (import.meta.env.DEV) {
        const contentType = response.headers.get('content-type') || '';

        // Vite serves CSS as a JS module in dev -> "text/javascript"
        if (contentType.includes('javascript')) {
          const strippedData = importedCssText
            .replace(/export\s+default\s+/g, '')
            .trim();

          // safer than eval: parse the JS string literal manually
          if (
            (strippedData.startsWith('"') && strippedData.endsWith('"')) ||
            (strippedData.startsWith("'") && strippedData.endsWith("'"))
          ) {
            importedCssText = strippedData.slice(1, -1); // drop quotes
          } else {
            // fallback if it doesnâ€™t look like a string
            importedCssText = eval(strippedData);
          }
        }
      }
      const importedCssAst = parser.parse(importedCssText);
      const embeddedFontFaceCss = fontFaceExtractor.stringify(importedCssAst);
      if (embeddedFontFaceCss) {
        fontFaceRules.push(adjustFontFaceUrls(embeddedFontFaceCss, url));
      }
    }
  }
};

export const extractAndInjectFonts = async (
  rootOrCss: Node | string,
  id: string = 'vuetify-json-forms-fonts',
): Promise<void> => {
  try {
    let styleTag = document.querySelector<HTMLStyleElement>(
      `style[id="${id}"]`,
    );
    if (!styleTag) {
      styleTag = document.createElement('style');
      styleTag.id = id;
      document.head.appendChild(styleTag);
    }

    const fontFaceRules: string[] = [];

    if (typeof rootOrCss === 'string') {
      // Process raw CSS text
      await processCssText(rootOrCss, fontFaceRules);
    } else if (rootOrCss instanceof ShadowRoot && rootOrCss.hasChildNodes()) {
      // Process ShadowRoot <style> tags
      const stylePromises: Promise<void>[] = [];
      for (const node of rootOrCss.childNodes) {
        if (node instanceof HTMLStyleElement && node.textContent) {
          stylePromises.push(processStyleNode(node, fontFaceRules));
        }
      }
      await Promise.all(stylePromises);
    }

    styleTag.textContent = fontFaceRules.join('\n');
  } catch (e) {
    console.error('extractAndInjectFonts error:', e);
  }
};
