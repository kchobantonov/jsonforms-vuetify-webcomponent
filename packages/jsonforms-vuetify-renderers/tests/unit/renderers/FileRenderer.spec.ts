import { describe, it, expect, beforeEach } from 'vitest';
import { clearAllIds } from '@jsonforms/core';
import FileRenderer, {
  entry as fileControlRendererEntry,
} from '../../../src/renderers/FileRenderer.vue';
import { mountJsonForms } from '../util';

describe('FileRenderer.vue', () => {
  const renderers = [fileControlRendererEntry];

  const data = true;
  const schema = {
    $schema: 'http://json-schema.org/draft-07/schema#',
    type: 'string',
    title: 'Image File',
    format: 'binary',
    formatMaximum: '1048576',
    contentMediaType: 'image/*',
    description:
      'File with maximum size of 1MB encoded as data URI and including the file name',
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'file placeholder',
      showUnfocusedDescription: true,
      vuetify: {
        'v-file-input': {
          showSize: true,
        },
      },
    },
  };
  let wrapper: ReturnType<typeof mountJsonForms>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  it('check if child FileRenderer exists', () => {
    expect(wrapper.getComponent(FileRenderer));
  });

  it('renders a file', () => {
    expect(wrapper.find('input[type="file"]').exists()).toBe(true);
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('Image File');
  });

  it('should work with accept', () => {
    expect(wrapper.find('input').element.getAttribute('accept')).toBe(
      'image/*',
    );
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="file"]');
    await input.trigger('focus');
    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('file placeholder');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
