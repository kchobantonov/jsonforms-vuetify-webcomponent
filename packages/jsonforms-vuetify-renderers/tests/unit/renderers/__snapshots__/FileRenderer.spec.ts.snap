// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileRenderer.vue should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         description="File with maximum size of 1MB encoded as data URI and including the file name"
         errors="must be string"
         label="Image File"
    >
      <div class="v-input v-input--has-state theme--light v-text-field v-text-field--placeholder v-file-input error--text input">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend">
            <button type="button"
                    aria-label="prepend icon"
                    class="v-icon notranslate v-icon--link mdi mdi-paperclip theme--light error--text"
            >
            </button>
          </div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot">
              <label for="#-input"
                     class="v-label theme--light error--text"
                     style="left: 0px; position: absolute;"
              >
                Image File
              </label>
              <div class="v-file-input__text v-file-input__text--placeholder">
              </div>
              <input accept="image/*"
                     id="#-input"
                     type="file"
              >
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light error--text"
                 role="alert"
            >
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
                <div class="v-messages__message">
                  must be string
                </div>
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <div class="v-dialog__container">
      </div>
    </div>
  </div>
</div>
`;
